Invalid auth tokens fail confusingly

**Describe the bug**
When the saved auth token (`$HOME/.flank/UserToken`) is invalid for any reason, it is hard to understand why it is invalid. It also does not clean itself up.


**To Reproduce**
> Steps to reproduce the behavior:

1. Run `java -jar flank.jar auth login`
2. Login
3. Wait long enough for this token to expire (I don't know how long this is)
4. Run flank again on the same computer

**Expected behavior**
> A clear and concise description of what you expected to happen.

I don't have a clear idea of what expected behavior is.
Idea 1 - Detect that the token is invalid, throw an error, and tell the user to run `auth login` again.
Idea 2- Automatically delete the token and fall back to env var credentials (via `$GOOGLE_APPLICATION_CREDENTIALS`)
Idea 3 - Clear log statement showing where flank is trying to pull credentials from, so at least when they see auth related errors, they know where to look


**Details (please complete the following information):**
> Have you tested on the latest Flank snapshot?
>
> Post the output of ` flank --version`.
Actual behavior:

I believe I had an expired token. When I tried running flank, I saw the error:
```
Caused by: com.google.api.client.http.HttpResponseException: 400 Bad Request
POST https://oauth2.googleapis.com/token
{
  "error": "invalid_grant",
  "error_description": "Bad Request"
}
```

I was playing around with `$GOOGLE_APPLICATION_CREDENTIALS`, but I had to manually step through the code to realize that I had a locally saved token that was taking precedence over the env var.

**Additional context**
> Add any other context about the problem here.

--- a/test_runner/src/main/kotlin/ftl/args/ArgsHelper.kt
+++ b/test_runner/src/main/kotlin/ftl/args/ArgsHelper.kt
@@ -10,15 +10,14 @@ import com.google.cloud.storage.BucketInfo
 import com.google.cloud.storage.Storage
 import com.google.cloud.storage.StorageClass
 import com.google.cloud.storage.StorageOptions
-import ftl.args.ArgsHelper.convertToWindowsPath
 import ftl.args.IArgs.Companion.AVAILABLE_PHYSICAL_SHARD_COUNT_RANGE
 import ftl.args.yml.YamlObjectMapper
-import ftl.config.FtlConstants
 import ftl.config.FtlConstants.GCS_PREFIX
 import ftl.config.FtlConstants.JSON_FACTORY
-import ftl.config.FtlConstants.defaultCredentialPath
+import ftl.config.defaultCredentialPath
 import ftl.config.FtlConstants.isWindows
 import ftl.config.FtlConstants.useMock
+import ftl.config.credential
 import ftl.gc.GcStorage
 import ftl.gc.GcToolResults
 import ftl.reports.xml.model.JUnitTestResult
@@ -146,7 +145,7 @@ object ArgsHelper {
         if (bucket.startsWith("test-lab-")) return bucket
 
         val storage = StorageOptions.newBuilder()
-            .setCredentials(FtlConstants.credential)
+            .setCredentials(credential)
             .setProjectId(projectId)
             .build().service
         val bucketLabel = mapOf("flank" to "")

--- a/test_runner/src/main/kotlin/ftl/args/ValidateCommonArgs.kt
+++ b/test_runner/src/main/kotlin/ftl/args/ValidateCommonArgs.kt
@@ -2,6 +2,7 @@ package ftl.args
 
 import ftl.config.Device
 import ftl.config.FtlConstants
+import ftl.config.defaultCredentialPath
 import ftl.gc.GcStorage
 import ftl.reports.FullJUnitReport
 import ftl.reports.JUnitReport
@@ -30,7 +31,7 @@ private fun List<Device>.throwIfAnyMisspelt() =
 private fun CommonArgs.assertProjectId() {
     if (project.isBlank()) throw FlankConfigurationError(
         "The project is not set. Define GOOGLE_CLOUD_PROJECT, set project in flank.yml\n" +
-            "or save service account credential to ${FtlConstants.defaultCredentialPath}\n" +
+            "or save service account credential to $defaultCredentialPath\n" +
             " See https://github.com/GoogleCloudPlatform/google-cloud-java#specifying-a-project-id"
     )
 }

--- a/test_runner/src/main/kotlin/ftl/config/Credentials.kt
+++ b/test_runner/src/main/kotlin/ftl/config/Credentials.kt
@@ -1,0 +1,52 @@
+package ftl.config
+
+import com.google.api.client.http.GoogleApiLogger
+import com.google.api.client.http.HttpRequestInitializer
+import com.google.auth.oauth2.AccessToken
+import com.google.auth.oauth2.GoogleCredentials
+import com.google.auth.oauth2.ServiceAccountCredentials
+import ftl.config.FtlConstants.userHome
+import ftl.gc.UserAuth
+import ftl.http.HttpTimeoutIncrease
+import ftl.run.exception.FlankGeneralError
+import java.io.IOException
+import java.nio.file.Path
+import java.nio.file.Paths
+import java.util.Date
+
+val defaultCredentialPath: Path by lazy {
+    Paths.get(userHome, ".config/gcloud/application_default_credentials.json")
+}
+
+val credential: GoogleCredentials by lazy {
+    when {
+        FtlConstants.useMock -> GoogleCredentials.create(AccessToken("mock", Date()))
+        UserAuth.exists() -> UserAuth.load()
+        else -> runCatching {
+            GoogleApiLogger.silenceComputeEngine()
+            ServiceAccountCredentials.getApplicationDefault()
+        }.getOrElse {
+            if (FtlConstants.isWindows) loadGoogleAccountCredentials()
+            else throw FlankGeneralError("Error: Failed to read service account credential.\n${it.message}")
+        }
+    }
+}
+
+private fun loadGoogleAccountCredentials(): GoogleCredentials = try {
+    GoogleCredentials.fromStream(defaultCredentialPath.toFile().inputStream())
+} catch (e: IOException) {
+    throw FlankGeneralError("Error: Failed to read service account credential.\n${e.message}")
+}
+
+val httpCredential: HttpRequestInitializer by lazy {
+    if (FtlConstants.useMock) {
+        HttpRequestInitializer {}
+    } else {
+        // Authenticate with https://github.com/googleapis/google-auth-library-java
+        // Scope is required.
+        // https://developers.google.com/identity/protocols/googlescopes
+        // https://developers.google.com/identity/protocols/application-default-credentials
+        // https://cloud.google.com/sdk/gcloud/reference/alpha/compute/instances/set-scopes
+        HttpTimeoutIncrease(credential.createScoped(listOf("https://www.googleapis.com/auth/cloud-platform")))
+    }
+}

--- a/test_runner/src/main/kotlin/ftl/config/FtlConstants.kt
+++ b/test_runner/src/main/kotlin/ftl/config/FtlConstants.kt
@@ -3,31 +3,24 @@ package ftl.config
 import com.bugsnag.Bugsnag
 import com.google.api.client.googleapis.javanet.GoogleNetHttpTransport
 import com.google.api.client.googleapis.util.Utils
-import com.google.api.client.http.GoogleApiLogger
-import com.google.api.client.http.HttpRequestInitializer
 import com.google.api.client.http.javanet.NetHttpTransport
 import com.google.api.client.json.JsonFactory
-import com.google.auth.oauth2.AccessToken
-import com.google.auth.oauth2.GoogleCredentials
-import com.google.auth.oauth2.ServiceAccountCredentials
 import ftl.args.AndroidArgs
 import ftl.args.IArgs
 import ftl.args.IosArgs
-import ftl.gc.UserAuth
-import ftl.http.HttpTimeoutIncrease
 import ftl.util.BugsnagInitHelper.initBugsnag
 import ftl.run.exception.FlankConfigurationError
 import ftl.run.exception.FlankGeneralError
 import ftl.util.readRevision
-import java.io.IOException
-import java.nio.file.Path
-import java.nio.file.Paths
-import java.util.Date
 
 object FtlConstants {
     var useMock = false
 
     private val osName = System.getProperty("os.name")?.toLowerCase() ?: ""
+
+    val userHome: String by lazy {
+        if (isWindows) System.getenv("HOMEPATH") else System.getProperty("user.home")
+    }
 
     val isMacOS: Boolean by lazy {
         val isMacOS = osName.indexOf("mac") >= 0
@@ -65,44 +58,6 @@ object FtlConstants {
             return@lazy GoogleNetHttpTransport.newTrustedTransport()
         } catch (e: Exception) {
             throw FlankGeneralError(e)
-        }
-    }
-
-    val defaultCredentialPath: Path by lazy {
-        val homePath = if (isWindows) System.getenv("HOMEPATH") else System.getProperty("user.home")
-        Paths.get(homePath, ".config/gcloud/application_default_credentials.json")
-    }
-
-    val credential: GoogleCredentials by lazy {
-        when {
-            useMock -> GoogleCredentials.create(AccessToken("mock", Date()))
-            UserAuth.exists() -> UserAuth.load()
-            else -> runCatching {
-                GoogleApiLogger.silenceComputeEngine()
-                ServiceAccountCredentials.getApplicationDefault()
-            }.getOrElse {
-                if (isWindows) loadGoogleAccountCredentials()
-                else throw FlankGeneralError("Error: Failed to read service account credential.\n${it.message}")
-            }
-        }
-    }
-
-    private fun loadGoogleAccountCredentials(): GoogleCredentials = try {
-        GoogleCredentials.fromStream(defaultCredentialPath.toFile().inputStream())
-    } catch (e: IOException) {
-        throw FlankGeneralError("Error: Failed to read service account credential.\n${e.message}")
-    }
-
-    val httpCredential: HttpRequestInitializer by lazy {
-        if (useMock) {
-            HttpRequestInitializer {}
-        } else {
-            // Authenticate with https://github.com/googleapis/google-auth-library-java
-            // Scope is required.
-            // https://developers.google.com/identity/protocols/googlescopes
-            // https://developers.google.com/identity/protocols/application-default-credentials
-            // https://cloud.google.com/sdk/gcloud/reference/alpha/compute/instances/set-scopes
-            HttpTimeoutIncrease(credential.createScoped(listOf("https://www.googleapis.com/auth/cloud-platform")))
         }
     }
 

--- a/test_runner/src/main/kotlin/ftl/gc/GcStorage.kt
+++ b/test_runner/src/main/kotlin/ftl/gc/GcStorage.kt
@@ -12,7 +12,7 @@ import ftl.args.IArgs
 import ftl.args.IosArgs
 import ftl.config.FtlConstants
 import ftl.config.FtlConstants.GCS_PREFIX
-import ftl.gc.GcStorage.dropLeadingSlash
+import ftl.config.credential
 import ftl.reports.xml.model.JUnitTestResult
 import ftl.reports.xml.parseAllSuitesXml
 import ftl.reports.xml.xmlToString
@@ -34,7 +34,7 @@ object GcStorage {
     val storageOptions: StorageOptions by lazy {
         val builder = StorageOptions.newBuilder()
         if (FtlConstants.useMock) builder.setHost(FtlConstants.localhost)
-        builder.setCredentials(FtlConstants.credential)
+        builder.setCredentials(credential)
 
         // The oauth lib for user auth needs to be replaced
         // https://github.com/Flank/flank/issues/464#issuecomment-455227703

--- a/test_runner/src/main/kotlin/ftl/gc/GcTesting.kt
+++ b/test_runner/src/main/kotlin/ftl/gc/GcTesting.kt
@@ -4,7 +4,7 @@ import com.google.api.services.testing.Testing
 import ftl.config.FtlConstants
 import ftl.config.FtlConstants.JSON_FACTORY
 import ftl.config.FtlConstants.applicationName
-import ftl.config.FtlConstants.httpCredential
+import ftl.config.httpCredential
 import ftl.config.FtlConstants.httpTransport
 import ftl.http.executeWithRetry
 

--- a/test_runner/src/main/kotlin/ftl/gc/GcToolResults.kt
+++ b/test_runner/src/main/kotlin/ftl/gc/GcToolResults.kt
@@ -20,6 +20,7 @@ import ftl.config.httpCredential
 import ftl.config.FtlConstants.httpTransport
 import ftl.http.executeWithRetry
 import ftl.run.exception.FTLProjectError
+import ftl.run.exception.FailureToken
 import ftl.run.exception.FlankGeneralError
 import ftl.run.exception.PermissionDenied
 import ftl.run.exception.ProjectNotFound
@@ -129,6 +130,7 @@ object GcToolResults {
         when (ftlProjectError) {
             is PermissionDenied -> throw FlankGeneralError(permissionDeniedErrorMessage(projectId, ftlProjectError.message))
             is ProjectNotFound -> throw FlankGeneralError(projectNotFoundErrorMessage(projectId, ftlProjectError.message))
+            is FailureToken -> UserAuth.throwAuthenticationError()
         }
     }
 

--- a/test_runner/src/main/kotlin/ftl/gc/GcToolResults.kt
+++ b/test_runner/src/main/kotlin/ftl/gc/GcToolResults.kt
@@ -16,7 +16,7 @@ import ftl.args.IArgs
 import ftl.config.FtlConstants
 import ftl.config.FtlConstants.JSON_FACTORY
 import ftl.config.FtlConstants.applicationName
-import ftl.config.FtlConstants.httpCredential
+import ftl.config.httpCredential
 import ftl.config.FtlConstants.httpTransport
 import ftl.http.executeWithRetry
 import ftl.run.exception.FTLProjectError

--- a/test_runner/src/main/kotlin/ftl/gc/UserAuth.kt
+++ b/test_runner/src/main/kotlin/ftl/gc/UserAuth.kt
@@ -1,10 +1,13 @@
 package ftl.gc
 
+import com.google.auth.oauth2.AccessToken
 import com.google.auth.oauth2.ClientId
 import com.google.auth.oauth2.MemoryTokensStorage
 import com.google.auth.oauth2.UserAuthorizer
 import com.google.auth.oauth2.UserCredentials
 import ftl.config.FtlConstants
+import ftl.config.FtlConstants.userHome
+import ftl.run.exception.FlankGeneralError
 import io.ktor.application.call
 import io.ktor.response.respondText
 import io.ktor.routing.get
@@ -18,21 +21,37 @@ import java.io.FileOutputStream
 import java.io.ObjectInputStream
 import java.io.ObjectOutputStream
 import java.net.URI
+import java.nio.file.Files
 import java.nio.file.Path
 import java.nio.file.Paths
+import java.util.Date
 
 class UserAuth {
 
     companion object {
-        private val home = System.getProperty("user.home")!!
-        private val dotFlank = Paths.get(home, ".flank/")
+        private val dotFlank = Paths.get(userHome, ".flank")
         val userToken: Path = Paths.get(dotFlank.toString(), "UserToken")
 
-        fun exists() = userToken.toFile().exists()
-        fun load(): UserCredentials {
-            return ObjectInputStream(FileInputStream(userToken.toFile())).use {
-                it.readObject() as UserCredentials
-            }
+        fun exists() = Files.exists(userToken)
+        fun load(): UserCredentials = readCredentialsOrThrow()
+
+        private fun readCredentialsOrThrow(): UserCredentials = runCatching {
+            ObjectInputStream(FileInputStream(userToken.toFile())).readObject() as UserCredentials
+        }.getOrElse {
+            throwAuthenticationError()
+        }.also {
+            ObjectOutputStream(FileOutputStream(userToken.toFile())).writeObject(
+                it.toBuilder().setAccessToken(AccessToken(it.accessToken.tokenValue, Date(0L))).build()
+            )
+        }
+
+        fun throwAuthenticationError(): Nothing {
+            Files.delete(userToken)
+            throw FlankGeneralError(
+                "Could not load user authentication, please\n" +
+                    " - login again using command: flank auth login\n" +
+                    " - or try again to use The Application Default Credentials variable to login"
+            )
         }
     }
 
@@ -96,9 +115,7 @@ class UserAuth {
         val userCredential = authorizer.getCredentials(userId)
 
         dotFlank.toFile().mkdirs()
-        ObjectOutputStream(FileOutputStream(userToken.toFile())).use {
-            it.writeObject(userCredential)
-        }
+        ObjectOutputStream(FileOutputStream(userToken.toFile())).writeObject(userCredential)
 
         println()
         println("User token saved to $userToken")

--- a/test_runner/src/main/kotlin/ftl/gc/UserAuth.kt
+++ b/test_runner/src/main/kotlin/ftl/gc/UserAuth.kt
@@ -1,6 +1,5 @@
 package ftl.gc
 
-import com.google.auth.oauth2.AccessToken
 import com.google.auth.oauth2.ClientId
 import com.google.auth.oauth2.MemoryTokensStorage
 import com.google.auth.oauth2.UserAuthorizer
@@ -16,6 +15,7 @@ import io.ktor.server.engine.embeddedServer
 import io.ktor.server.netty.Netty
 import kotlinx.coroutines.delay
 import kotlinx.coroutines.runBlocking
+import java.io.File
 import java.io.FileInputStream
 import java.io.FileOutputStream
 import java.io.ObjectInputStream
@@ -24,13 +24,13 @@ import java.net.URI
 import java.nio.file.Files
 import java.nio.file.Path
 import java.nio.file.Paths
-import java.util.Date
 
 class UserAuth {
 
     companion object {
         private val dotFlank = Paths.get(userHome, ".flank")
-        val userToken: Path = Paths.get(dotFlank.toString(), "UserToken")
+        val userToken: Path = if (FtlConstants.useMock) File.createTempFile("test_", ".Token").toPath()
+        else Paths.get(dotFlank.toString(), "UserToken")
 
         fun exists() = Files.exists(userToken)
         fun load(): UserCredentials = readCredentialsOrThrow()
@@ -39,10 +39,6 @@ class UserAuth {
             ObjectInputStream(FileInputStream(userToken.toFile())).readObject() as UserCredentials
         }.getOrElse {
             throwAuthenticationError()
-        }.also {
-            ObjectOutputStream(FileOutputStream(userToken.toFile())).writeObject(
-                it.toBuilder().setAccessToken(AccessToken(it.accessToken.tokenValue, Date(0L))).build()
-            )
         }
 
         fun throwAuthenticationError(): Nothing {

--- a/test_runner/src/main/kotlin/ftl/http/ExecuteWithRetry.kt
+++ b/test_runner/src/main/kotlin/ftl/http/ExecuteWithRetry.kt
@@ -3,6 +3,7 @@ package ftl.http
 import com.google.api.client.googleapis.services.json.AbstractGoogleJsonClientRequest
 import com.google.api.client.http.HttpResponseException
 import ftl.config.FtlConstants
+import ftl.gc.UserAuth
 import ftl.run.exception.PermissionDenied
 import ftl.run.exception.ProjectNotFound
 import kotlinx.coroutines.delay
@@ -29,6 +30,7 @@ private inline fun <T> withRetry(crossinline block: () -> T): T = runBlocking {
             if (err is HttpResponseException) {
                 // we want to handle some FTL errors with special care
                 when (err.statusCode) {
+                    400 -> if (err.containsBadTokenMessage()) UserAuth.throwAuthenticationError()
                     429 -> return@repeat
                     403 -> throw PermissionDenied(err)
                     404 -> throw ProjectNotFound(err)
@@ -41,3 +43,7 @@ private inline fun <T> withRetry(crossinline block: () -> T): T = runBlocking {
 }
 
 private class FlankGoogleApiError(exception: Throwable) : Error(exception)
+
+private fun HttpResponseException.containsBadTokenMessage() = message?.let {
+    it.contains("\"error\": \"invalid_grant\"") && it.contains("https://oauth2.googleapis.com/token")
+} ?: false

--- a/test_runner/src/main/kotlin/ftl/http/ExecuteWithRetry.kt
+++ b/test_runner/src/main/kotlin/ftl/http/ExecuteWithRetry.kt
@@ -3,7 +3,7 @@ package ftl.http
 import com.google.api.client.googleapis.services.json.AbstractGoogleJsonClientRequest
 import com.google.api.client.http.HttpResponseException
 import ftl.config.FtlConstants
-import ftl.gc.UserAuth
+import ftl.run.exception.FailureToken
 import ftl.run.exception.PermissionDenied
 import ftl.run.exception.ProjectNotFound
 import kotlinx.coroutines.delay
@@ -30,7 +30,7 @@ private inline fun <T> withRetry(crossinline block: () -> T): T = runBlocking {
             if (err is HttpResponseException) {
                 // we want to handle some FTL errors with special care
                 when (err.statusCode) {
-                    400 -> if (err.containsBadTokenMessage()) UserAuth.throwAuthenticationError()
+                    400 -> if (err.containsBadTokenMessage()) throw FailureToken(err) else return@repeat
                     429 -> return@repeat
                     403 -> throw PermissionDenied(err)
                     404 -> throw ProjectNotFound(err)

--- a/test_runner/src/main/kotlin/ftl/run/exception/FlankException.kt
+++ b/test_runner/src/main/kotlin/ftl/run/exception/FlankException.kt
@@ -80,6 +80,7 @@ class FlankGeneralError : FlankException {
 sealed class FTLProjectError(exc: IOException) : FlankException("Caused by: $exc")
 class PermissionDenied(exc: IOException) : FTLProjectError(exc)
 class ProjectNotFound(exc: IOException) : FTLProjectError(exc)
+class FailureToken(exc: IOException) : FTLProjectError(exc)
 
 /**
  * The test environment for this test execution is not supported because of incompatible test dimensions. This error might occur if the selected Android API level is not supported by the selected device type.

--- a/test_runner/src/test/kotlin/ftl/gc/GcToolResultsTest.kt
+++ b/test_runner/src/test/kotlin/ftl/gc/GcToolResultsTest.kt
@@ -6,6 +6,7 @@ import com.google.api.services.testing.model.ToolResultsHistory
 import com.google.common.truth.Truth.assertThat
 import ftl.args.AndroidArgs
 import ftl.config.FtlConstants
+import ftl.run.exception.FailureToken
 import ftl.test.util.FlankTestRunner
 import ftl.test.util.TestHelper.getThrowable
 import ftl.run.exception.FlankGeneralError
@@ -90,7 +91,9 @@ class GcToolResultsTest {
             }
             """.trimIndent()
             val mockJSonException = GoogleJsonResponseException(exceptionBuilder, null)
-            every { GcToolResults.service.Projects().initializeSettings(projectId) } throws PermissionDenied(mockJSonException)
+            every { GcToolResults.service.Projects().initializeSettings(projectId) } throws PermissionDenied(
+                mockJSonException
+            )
             val exception = getThrowable { GcToolResults.getDefaultBucket(projectId) }
             assertEquals(expected, exception.message)
         }
@@ -100,7 +103,9 @@ class GcToolResultsTest {
     fun `getDefaultBucket on PermissionDenied error should throw FlankGeneralError`() {
         mockkObject(GcToolResults) {
             every { GcToolResults.service.applicationName } returns projectId
-            every { GcToolResults.service.Projects().initializeSettings(projectId) } throws PermissionDenied(IOException())
+            every {
+                GcToolResults.service.Projects().initializeSettings(projectId)
+            } throws PermissionDenied(IOException())
             GcToolResults.getDefaultBucket(projectId)
         }
     }
@@ -109,7 +114,9 @@ class GcToolResultsTest {
     fun `getDefaultBucket on ProjectNotFound error should throw FlankGeneralError`() {
         mockkObject(GcToolResults) {
             every { GcToolResults.service.applicationName } returns projectId
-            every { GcToolResults.service.Projects().initializeSettings(projectId) } throws ProjectNotFound(IOException())
+            every {
+                GcToolResults.service.Projects().initializeSettings(projectId)
+            } throws ProjectNotFound(IOException())
             GcToolResults.getDefaultBucket(projectId)
         }
     }
@@ -149,8 +156,39 @@ class GcToolResultsTest {
             }
             """.trimIndent()
             val mockJSonException = GoogleJsonResponseException(exceptionBuilder, null)
-            every { GcToolResults.service.Projects().initializeSettings(projectId) } throws ProjectNotFound(mockJSonException)
+            every { GcToolResults.service.Projects().initializeSettings(projectId) } throws ProjectNotFound(
+                mockJSonException
+            )
             val exception = getThrowable { GcToolResults.getDefaultBucket(projectId) }
+            assertEquals(expected, exception.message)
+        }
+    }
+
+    @Test
+    fun `getDefaultBucket on 400 with authentication error should throw exception with specific message`() {
+        val expected = """
+Could not load user authentication, please
+ - login again using command: flank auth login
+ - or try again to use The Application Default Credentials variable to login
+            """.trimIndent()
+        mockkObject(GcToolResults) {
+            every { GcToolResults.service.applicationName } returns FtlConstants.applicationName
+
+            val exceptionBuilder = mockk<HttpResponseException.Builder>()
+            every { exceptionBuilder.message } returns """
+Caused by: com.google.api.client.http.HttpResponseException: 400 Bad Request
+POST https://oauth2.googleapis.com/token
+{
+  "error": "invalid_grant",
+  "error_description": "Bad Request"
+}
+            """.trimIndent()
+            val mockJSonException = GoogleJsonResponseException(exceptionBuilder, null)
+            every { GcToolResults.service.Projects().initializeSettings(projectId) } throws FailureToken(
+                mockJSonException
+            )
+            val exception = getThrowable { GcToolResults.getDefaultBucket(projectId) }
+            assertThat(exception).isInstanceOf(FlankGeneralError::class.java)
             assertEquals(expected, exception.message)
         }
     }

--- a/test_runner/src/test/kotlin/ftl/gc/UserAuthTest.kt
+++ b/test_runner/src/test/kotlin/ftl/gc/UserAuthTest.kt
@@ -1,0 +1,32 @@
+package ftl.gc
+
+import com.google.common.truth.Truth.assertThat
+import ftl.gc.UserAuth.Companion.userToken
+import ftl.run.exception.FlankGeneralError
+import ftl.test.util.TestHelper.getThrowable
+import io.mockk.every
+import io.mockk.mockkObject
+import org.junit.Test
+import java.io.File
+
+class UserAuthTest {
+
+    @Test
+    fun `should delete file and throw exception with correct message for incorrect file`() {
+        val tempFile = File.createTempFile("test_", ".Token")
+        val expectedMessage = """
+Could not load user authentication, please
+ - login again using command: flank auth login
+ - or try again to use The Application Default Credentials variable to login
+            """.trimIndent()
+        mockkObject(UserAuth.Companion) {
+            every { userToken } returns tempFile.toPath()
+
+            getThrowable { UserAuth.load() }.run {
+                assertThat(this).isInstanceOf(FlankGeneralError::class.java)
+                assertThat(message).isEqualTo(expectedMessage)
+                assertThat(tempFile.exists()).isFalse()
+            }
+        }
+    }
+}

